linked_list/linked_list.py

```diff
@@ -9,109 +9,205 @@ def __init__(self, value, next_node = None):
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def delete(self, value):
-        pass
+        if self.size == 0:
+            return None
+
+        current = self.head
+        if current.value == value:
+            self.head = current.next
+
+        while current.next:
+            if current.next.value == value:
+                current.next = current.next.next
+            else:
+                current = current.next
+        self.size -= 1
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def visit(self):
         helper_list = []
         current = self.head
 
         while current:
             helper_list.append(str(current.value))
             current = current.next
-        
+
         print(", ".join(helper_list))
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def reverse(self):
-        pass
-  
+        if self.size == 0:
+            return None
+
+        previous = None
+        current = self.head
+
+        while current:
+            next = current.next
+            current.next = previous
+            previous = current
+            current = next
+        self.head = previous
+
     ## Advanced/ Exercises
     # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_middle_value(self):
-        pass
+        middle_index = (self.size // 2)
+        return self.get_at_index(middle_index)
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_nth_from_end(self, n):
-        pass
+        index = (self.size - n) - 1
+        return self.get_at_index(index)
+
+    # returns the node at a given index, rather than the value
+    def get_node_at_index(self, index):
+        if (self.size + 1) < index:
+            return None
+
+        counter = 0
+        current = self.head
+        while current:
+            if counter == index:
+                return current
+            current = current.next
+            counter += 1
 
     # checks if the linked list has a cycle. A cycle exists if any node in the
     # linked list links to a node already visited.
     # returns true if a cycle is found, false otherwise.
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def has_cycle(self):

```

Comment:
⚠️ This only works if the loop goes back to the head, which isn't always the case.

Reply:
Thanks Chris! I misunderstood the definition of a cycle. Looking at the code comments and tests, I thought it meant that the tail had to be connected to the head (I thought of an ouroboros). I'll think of some other ways I could account for the loop going back to any node, rather than just the head.